coredb-operator/src/statefulset.rs

```diff
@@ -309,12 +314,35 @@ pub async fn reconcile_sts(cdb: &CoreDB, ctx: Arc<Context>) -> Result<(), Error>
     let sts_api: Api<StatefulSet> = Api::namespaced(client, &sts.clone().metadata.namespace.unwrap());
     // If cdb is running and storage is resized in CoreDB Custom Resource then follow steps to resize PVC
     // Reference Article: https://itnext.io/resizing-statefulset-persistent-volumes-with-zero-downtime-916ebc65b1d4
-    if cdb.status.is_some()
-        && cdb.status.clone().unwrap().running
-        && cdb.status.clone().unwrap().storage != cdb.spec.storage
-    {
-        delete_sts(ctx.client.clone(), &sts).await;
-        update_pvc(ctx.client.clone(), &sts, cdb).await;
+
+    // determine pvcs that changed
+    let pvcs_to_update = match cdb.status.is_some() && cdb.status.clone().unwrap().running {
+        true => {
+            let mut pvcs_to_update = Vec::new();
+            if cdb.status.clone().unwrap().storage != cdb.spec.storage {
+                pvcs_to_update.push(("data".to_string(), cdb.spec.storage.clone()));
+            }
+            if cdb.status.clone().unwrap().sharedirStorage != cdb.spec.sharedirStorage {
+                pvcs_to_update.push(("sharedir".to_string(), cdb.spec.sharedirStorage.clone()));
+            }
+            if cdb.status.clone().unwrap().pkglibdirStorage != cdb.spec.pkglibdirStorage {
+                pvcs_to_update.push(("pkglibdir".to_string(), cdb.spec.pkglibdirStorage.clone()));
+            }
+            pvcs_to_update
+        }
+        false => Vec::new(),
+    };
+    if !pvcs_to_update.is_empty() {
+        let sts_name = sts.clone().metadata.name.unwrap();
+        let sts_namespace = sts.clone().metadata.namespace.unwrap();
+
+        // why do delete first, then update?
+        delete_sts_no_cascade(&sts_api, &sts_name).await;
+
+        let pvc_api: Api<PersistentVolumeClaim> = Api::namespaced(ctx.client.clone(), &sts_namespace);
+        for (pvc_name, qty) in pvcs_to_update {
+            update_pvc(&pvc_api, &sts_name, &pvc_name, qty).await;
+        }
     }
```

Comment:
@sjmiller609 - can you remind me why we first delete the sts, then update the pvc?

Reply:
deleting the sts without cascade prevents the sts from trying to manage the pvcs it owns. Then increase the pvcs to match what they will be in the new sts, then create new sts which discovers the existing pvcs and adopts them. It might be possible to update the pvcs, then delete sts, then recreate sts, but I'm not sure if that order works.

Reply:
Updating PVC's in an already defined STS is immutable and wont update.  So you are forced to delete the sts first then have the operator recreate the sts.